Web Automation: implement Set Storage Access endpoint

burg opened the pull request:
#### 19e192754fe6cf1083b9a87bd1bb532b2c7cde11
<pre>
Web Automation: implement Set Storage Access endpoint
<a href="https://bugs.webkit.org/show_bug.cgi?id=297368">https://bugs.webkit.org/show_bug.cgi?id=297368</a>
<a href="https://rdar.apple.com/158263193">rdar://158263193</a>

Reviewed by Charlie Wolfe.

Add new endpoints for setting storage access permission state and
granting storage access to embedded frames for specific origins.

These methods will be called via Automation protocol to implement the
WebDriver Classic endpoints in safaridriver. In a future patch, the
WebAutomationSession methods will be used by the WebDriver BiDi permissions agent.

* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::setStorageAccessPermissionState):
(WebKit::WebAutomationSession::grantStorageAccess):

[WebDriver BiDi] add support for permissions.setPermission command
<a href="https://bugs.webkit.org/show_bug.cgi?id=265621">https://bugs.webkit.org/show_bug.cgi?id=265621</a>
<a href="https://rdar.apple.com/119346759">rdar://119346759</a>

Reviewed by Charlie Wolfe.

Implement the remote end steps for Set Permission.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Automation/BidiPermissionsAgent.cpp: Added.
(WebKit::BidiPermissionsAgent::BidiPermissionsAgent):
(WebKit::allPageProxiesFor):
(WebKit::BidiPermissionsAgent::setPermission):
Added. Defer to WebSiteDataStore and NetworkProcess for the actual implementation.

* Source/WebKit/UIProcess/Automation/BidiPermissionsAgent.h: Added.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
Add BidiPermissionAgent.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiPermissions.json: Added.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Add new domain and associated derived sources and project entries.

Canonical link: <a href="https://commits.webkit.org/299547@main">https://commits.webkit.org/299547@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0fdd1163e0c6f9976cd53e2f573c95a017d1bba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119339 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39026 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125589 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71410 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4880b38c-9051-4f8f-bb36-7d71609aaed3) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47607 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90681 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/af45cfcc-f628-4975-b499-8e9e5f66fe02) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122291 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31692 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106996 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71121 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/77b4fc46-48a8-4d28-af37-72ea1f299ab0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30732 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69236 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101148 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25296 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128584 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46257 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34995 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99244 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103195 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99029 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25178 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44485 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22495 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42824 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46120 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51820 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45585 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48935 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47272 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebKit/UIProcess/Automation/Automation.json:
```diff
@@ -909,6 +920,29 @@
+                { "name": "topFrameDomain", "type": "string", "description": "Origin whose storage is being accessed." },
+                { "name": "subFrameDomain", "type": "string", "description": "Origin of the subframe that should be granted non-partitioned storage access to the top frame's origin." }
+            ],
+            "async": true
```
based on the linked spec it seems like there should be a `blocked` parameter?  is there a reason why this doesn't have that?

burg replied:
Handled in new version.

dcrousso commented on Source/WebKit/UIProcess/Automation/BidiPermissionsAgent.cpp:
```diff
@@ -0,0 +1,104 @@
+    Vector<Ref<WebPageProxy>> pages;
+    for (Ref process : session.protectedProcessPool()->processes()) {
+        for (Ref page : process->pages()) {
+            if (!page->isControlledByAutomation())
+                continue;
+            pages.append(WTFMove(page));
```
wait why do we only want pages not controlled by automation?

the-chenergy replied:
This code skips pages not controlled by automation.

dcrousso replied:
oh oops i completely misread that 🤣

dcrousso commented on Source/WebKit/UIProcess/Automation/BidiPermissionsAgent.cpp:
```diff
@@ -0,0 +1,104 @@
+                continue;
+
+            bool granted = state == Inspector::Protocol::BidiPermissions::PermissionState::Granted;
+            page->protectedWebsiteDataStore()->setStorageAccessPermissionForTesting(granted, page->identifier(), topFrameOrigin.string(), (embeddedOriginIsWildcard ? topFrameOrigin : embeddedOrigin).string(), [aggregator] () { });
```
Aside: unfortunate that this doesn't accept `RegistrableDomain` when it converts the arguments into one anyways :(

the-chenergy replied:
```suggestion
            page->protectedWebsiteDataStore()->setStorageAccessPermissionForTesting(granted, page->identifier(), topFrameOrigin.string(), (embeddedOriginIsWildcard ? topFrameOrigin : embeddedOrigin).string(), [aggregator]() { });
```

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -1789,6 +1789,26 @@ CommandResult<void> WebAutomationSession::generateTestReport(const String& brows
+    auto page = webPageProxyForHandle(browsingContextHandle);
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!page, WindowNotFound);
+
+    page->protectedWebsiteDataStore()->setStorageAccessPermissionForTesting(state == Inspector::Protocol::Automation::PermissionState::Granted, page->identifier(), topFrameOrigin, subFrameOrigin, [protectedThis = Ref { *this }, callback = WTFMove(callback)]() {
```
what about `PermissionState::Prompt`?

burg replied:
Handled in new version.

the-chenergy commented on Source/WebKit/UIProcess/Automation/BidiPermissionsAgent.cpp:
```diff
@@ -0,0 +1,104 @@
+WTF_MAKE_TZONE_ALLOCATED_IMPL(BidiPermissionsAgent);
+
+BidiPermissionsAgent::BidiPermissionsAgent(WebAutomationSession& session, BackendDispatcher& backendDispatcher)
+    : m_session(session)
```
Do we not need a domain dispatcher, similar to what I see in other agents?
```suggestion
    : m_session(session)
    , m_permissionDomainDispatcher(BidiPermissionBackendDispatcher::create(backendDispatcher, this))
```

burg replied:
If it's not used, then I don't think that it's necessary to allocate it.

the-chenergy commented on Source/WebKit/UIProcess/Automation/BidiPermissionsAgent.cpp:
```diff
@@ -0,0 +1,104 @@
+        bool embeddedOriginIsWildcard = subFrameURLString == "*"_s;
+        auto embeddedOrigin = RegistrableDomain { URL { subFrameURLString } };
+
+        auto aggregator = WTF::CallbackAggregator::create([callback = WTFMove(callback)]() mutable {
```
Should either this lambda or the one below also capture a `protectedThis = Ref { *this }`?

burg replied:
I wrote that before, but I removed it because the code doesn't actually use `this` in the callback. We have captured a ref to `backendDispatcher` in the generated lambda glue code that calls into the agent method.

```
m_agent->setPermission(in_descriptor.releaseNonNull(), in_origin, *in_state, in_opt_userContext, [backendDispatcher = m_backendDispatcher.copyRef(), 
...
```

So it shouldn't be necessary to protect `this`, as BackendDispatcher and FrontendRounter will gracefully stop sending responses when there's no more frontends connected. In the WebDriver BiDi case, the BidiProcessor owns both frontend and backend dispatcher. So, if a callback outlasts the processor itself, sendResponse will dead end at the frontend router that doesn't have any connections.

I did a deep dive on the refcounting situation and i don't see the need to protect the agent as its owned by the processor. Plus, in the WebDriver BiDi case, the processor is used as a fake "connection" so that it can forward events and command responses through to `WebAutomationSession::sendBidiMessage`.

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/protocol/BidiPermissions.json:
```diff
@@ -0,0 +1,44 @@
+            "description": "Simulates user modification of a PermissionDescriptor's permission state.",
+            "spec": "https://www.w3.org/TR/permissions/#webdriver-bidi-command-permissions-setPermission",
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/permissions/set",
+            "async": true,
```
Nit: async usually at the bottom

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/protocol/BidiPermissions.json:
```diff
@@ -0,0 +1,44 @@
+            "name": "setPermission",
+            "description": "Simulates user modification of a PermissionDescriptor's permission state.",
+            "spec": "https://www.w3.org/TR/permissions/#webdriver-bidi-command-permissions-setPermission",
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/permissions/set",
```
Link seems buggy ? I don't see a permissions bidi suite

burg replied:
Good catch, this is the correct link:
https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/external/permissions/set_permission

it's in a separate `external` directory because it's defined externally from the main bidi spec.

burg committed:
Web Automation: implement Set Storage Access endpoint
https://bugs.webkit.org/show_bug.cgi?id=297368
rdar://158263193

Reviewed by Charlie Wolfe.

Add new endpoints for setting storage access permission state and
granting storage access to embedded frames for specific origins.

These methods will be called via Automation protocol to implement the
WebDriver Classic endpoints in safaridriver. In a future patch, the
WebAutomationSession methods will be used by the WebDriver BiDi permissions agent.

* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::setStorageAccessPermissionState):
(WebKit::WebAutomationSession::grantStorageAccess):

[WebDriver BiDi] add support for permissions.setPermission command
https://bugs.webkit.org/show_bug.cgi?id=265621
rdar://119346759

Reviewed by Charlie Wolfe.

Implement the remote end steps for Set Permission.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Automation/BidiPermissionsAgent.cpp: Added.
(WebKit::BidiPermissionsAgent::BidiPermissionsAgent):
(WebKit::allPageProxiesFor):
(WebKit::BidiPermissionsAgent::setPermission):
Added. Defer to WebSiteDataStore and NetworkProcess for the actual implementation.

* Source/WebKit/UIProcess/Automation/BidiPermissionsAgent.h: Added.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
Add BidiPermissionAgent.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiPermissions.json: Added.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Add new domain and associated derived sources and project entries.

Canonical link: https://commits.webkit.org/299547@main
